General mods (#166)
decentralized-identity/decentralized-web-node: spec/spec.md

  - If the [Message](#messages) has data associated with it, passed directly via the `data` property of the [Message](#messages) object or through a channel external to the message object, the `descriptor` object ****MUST**** contain a `dataFormat` property, and its value ****MUST**** be a string that corresponds with a registered [IANA Media Type](https://www.iana.org/assignments/media-types/media-types.xhtml) data format (the most common being plain JSON, which is indicated by setting the value of the `dataFormat` property to `application/json`), or one of the following format strings pending registration:
    - `application/vc+jwt` - the data is a JSON Web Token (JWT) [[spec:rfc7519]] formatted variant of a [W3C Verifiable Credential](https://www.w3.org/TR/vc-data-model/#json-web-token).
    - `application/vc+ldp` - the data is a JSON-LD formatted [W3C Verifiable Credential](https://www.w3.org/TR/vc-data-model).

::: note
Individual Interface methods may describe additional properties that the `descriptor` object ****MUST**** or ****MAY**** contain, which are detailed in the [Interfaces](#interfaces) section of the specification.
:::

### Message Authorization

> Some messages may require authorization material for processing them in accordance with the permissions a [ref: DWeb Node] owner has specified. If a message requires authorization it ****MUST**** include an `authorization` property with a value that is a [spec:rfc7515]] Compact JSON Web Signature (JWS), constructed as follows:

Reviewer: ```suggestion
Some messages may require authorization material in order to process them in accordance with the permissions a [ref: DWeb Node] owner has specified. If a message requires authorization, it ****MUST**** include an `authorization` property with a value that is a [[spec:rfc7515]] Compact JSON Web Signature (JWS), constructed as follows:
```

        "threadId": "b6464162-84af-4aab-aff5-f1f8438dfc1e",
        "dataCid": CID(data),
        "dateCreated": 123456789,
        "dataFormat": "application/json"
      },
      "attestation": "43595vnwcn549h4wouberinciuqb3.sdfaf34fac34c5s45va5e.34qca5e5vs5r56brs65bs5",
        // {
        //   "protected": {
        //     "alg": "ES256K",
        //     "kid": "did:example:123#key-1"
        //   },
        //   "payload": CID(descriptor),
        //   "signature": Sign(protected + payload)
        // }
      "authorization": "5ev98h45ch8h758e95hw4wv9wv.598wh6v8n58v9h6w89trn5.f5v6eyw45657e6rutys5v6"
        // {
        //   "protected": {
        //     "alg": "ES256K",
        //     "kid": "did:example:456#key-1"
        //   },
        //   "payload": {
        //       "descriptorCid": "bafybeigdyrzt5sfp7udm7hu76uh7y26nf3efuylqabf3oclgtqy55fbzdi",
        //       "permissionsGrantCid": "baiv9sdhv87agdsvadhf7gh9h8jcw8e57hra7chcag7wh478gca78c7584g",
        //       "attestationCid": "baidfa98fhda8hsd9v87agdshf89ahsdf7va9sd98vahs98dhv9da8sh9ds"   
        //   },
        //   "signature": Sign(protected + payload)
        // }
    },
    {...}
  ]
}
```

> - The JWS ****MUST**** include a `protected` property, and its value ****must**** be an object composed of the following values:

Reviewer: ```suggestion
- The JWS ****MUST**** include a `protected` property, and its value ****MUST**** be an object composed of the following values:
```

  - The object ****MUST**** include an `alg` property, and its value ****MUST**** be the string representing the algorithm used to verify the signature (as defined by the [[spec:rfc7515]] JSON Web Signature specification).
  - The object ****MUST**** include a `kid` property, and its value ****MUST**** be a [DID URL](https://w3c.github.io/did-core/#example-a-unique-verification-method-in-a-did-document) string identifying the key to be used in verifying the signature.
- The JWS ****MUST**** include a `payload` property, and its value ****must**** be an object composed of the following values:

>   - The object ****MUST**** include an `descriptorCid` property, and its value ****MUST**** be the stringified [Version 1 CID](https://docs.ipfs.io/concepts/content-addressing/#identifier-formats) of the [DAG CBOR](https://github.com/ipld/specs/blob/master/block-layer/codecs/dag-cbor.md) encoded `descriptor` object.
>   - The object ****MUST**** include an `permissionsGrantCid` property, and its value ****MUST**** be the stringified [Version 1 CID](https://docs.ipfs.io/concepts/content-addressing/#identifier-formats) of the [DAG CBOR](https://github.com/ipld/specs/blob/master/block-layer/codecs/dag-cbor.md) encoded  [Permission Grant](#grant) being invoked.
>   - If attestation of an object is permitted the `payload` ****MAY**** include an `attestationCid` property, and its value ****MUST**** be the stringified [Version 1 CID](https://docs.ipfs.io/concepts/content-addressing/#identifier-formats) of the [DAG CBOR](https://github.com/ipld/specs/blob/master/block-layer/codecs/dag-cbor.md) encoded `attestation` string.

Reviewer: ```suggestion
  - The object ****MUST**** include a `descriptorCid` property, and its value ****MUST**** be the stringified [Version 1 CID](https://docs.ipfs.io/concepts/content-addressing/#identifier-formats) of the [DAG CBOR](https://github.com/ipld/specs/blob/master/block-layer/codecs/dag-cbor.md) encoded `descriptor` object.
  - The object ****MUST**** include a `permissionsGrantCid` property, and its value ****MUST**** be the stringified [Version 1 CID](https://docs.ipfs.io/concepts/content-addressing/#identifier-formats) of the [DAG CBOR](https://github.com/ipld/specs/blob/master/block-layer/codecs/dag-cbor.md) encoded  [Permission Grant](#grant) being invoked.
  - If attestation of an object is permitted, the `payload` ****MAY**** include an `attestationCid` property, the value of which ****MUST**** be the stringified [Version 1 CID](https://docs.ipfs.io/concepts/content-addressing/#identifier-formats) of the [DAG CBOR](https://github.com/ipld/specs/blob/master/block-layer/codecs/dag-cbor.md) encoded `attestation` string.
```

### Raw Data

> If there is no need or desire to sign or encrypt the content of a message (i.e. public repudiable data), the message `descriptor` object is the only property required in a [Message](#messages) (with any method-specific properties required). An optional `data` property may be passed at the [Message](#messages) level that contains the data associated with the message (when data is desired or required to be present for a given method invocation).

Reviewer: ```suggestion
If there is no need or desire to sign or encrypt the content of a message (i.e., public repudiable data), the message `descriptor` object is the only property required in a [Message](#messages) (along with any required method-specific properties). An optional `data` property may be passed at the [Message](#messages) level that contains the data associated with the message (when data is desired or required to be present for a given method invocation).
```

}
```

### Signed Data

> If the object is to be attested by a signer (e.g the Node owner via signature with their DID key), the object ****MUST**** contain the following additional properties to produce a [[spec:rfc7515]] Compact JSON Web Signature (JWS):

Reviewer: ```suggestion
If the object is to be attested by a signer (e.g., the Node owner, via signature with their DID key), the object ****MUST**** contain the following additional properties to produce a [[spec:rfc7515]] Compact JSON Web Signature (JWS):
```

    - The object ****MUST**** include a `protected` property, and its value ****must**** be an object composed of the following values:
        - The object ****MUST**** include an `alg` property, and its value ****MUST**** be the string representing the algorithm used to verify the signature (as defined by the [[spec:rfc7515]] JSON Web Signature specification).
        - The object ****MUST**** include a `kid` property, and its value ****MUST**** be a [DID URL](https://w3c.github.io/did-core/#example-a-unique-verification-method-in-a-did-document) string identifying the key to be used in verifying the signature.
    - The object ****MUST**** include a `signature` property, and its value ****must**** be a signature string produced by signing the `protected` and `payload` values, in accordance with the [[spec:rfc7515]] JSON Web Signature specification.

### Encrypted Data

> If the object is to be encrypted (e.g the Node owner encrypting their data to keep it private), the `descriptor` object ****MUST**** be constructed as follows:

Reviewer: ```suggestion
If the object is to be encrypted (e.g., the Node owner is encrypting their data to keep it private), the `descriptor` object ****MUST**** be constructed as follows:
```
